File: packages/swift/Sources/PylonRealtime/Replication.swift
import Foundation

/// Entity replication frames (`pylon_replication::frame` in Rust). A shard
/// that replicates entities sends each subscriber, per tick, despawns for
/// entities that left its view, spawns (full state) for entities that came
/// into view, and updates carrying only the axes that moved (as quantized
/// differences) and the components that changed or were removed.
/// `EntityTable` applies them.
public enum ReplicationError: Error, Equatable, CustomStringConvertible {
    case malformed(String)

    public var description: String {
        switch self {
        case .malformed(let why): return "replication frame: \(why)"
        }
    }
}

/// One entity as the client sees it.
public struct ReplicatedEntity: Sendable, Equatable {
    public let id: UInt64
    /// Quantized position; `position` is this times the frame precision.
    public var q: [Int64]
    public var position: (x: Double, y: Double, z: Double)
    /// Component id to bytes, in the game's own encoding.
    public var components: [UInt8: Data]

    public static func == (a: ReplicatedEntity, b: ReplicatedEntity) -> Bool {
        a.id == b.id && a.q == b.q && a.components == b.components
    }
}

/// What one frame did.
public struct ReplicationSummary: Sendable, Equatable {
    public var full = false
    public var spawned: [UInt64] = []
    public var updated: [UInt64] = []
    public var despawned: [UInt64] = []
}

/// The entities one subscriber has been told about. Apply every frame in
/// order; a frame that fails to apply means the connection is out of sync
/// and should be reopened (the server then sends a full frame).
public struct EntityTable: Sendable {
    public private(set) var entities: [UInt64: ReplicatedEntity] = [:]
    /// World units per quantization step, from the last frame.
    public private(set) var precision: Float = 0.01

    public init() {}

    public var count: Int { entities.count }

    public subscript(id: UInt64) -> ReplicatedEntity? { entities[id] }

    public mutating func clear() { entities.removeAll() }

    private static let maxCount: UInt64 = 1 << 24

    @discardableResult
    public mutating func apply(_ frame: Data) throws -> ReplicationSummary {
        var r = Reader(bytes: [UInt8](frame))
        let version = try r.u8()
        guard version == 1 else { throw ReplicationError.malformed("version \(version)") }
        let full = (try r.u8() & 1) != 0
        let precision = try r.f32()
        guard precision.isFinite, precision > 0 else { throw ReplicationError.malformed("bad precision") }
        if full { entities.removeAll() }
        self.precision = precision
        var summary = ReplicationSummary(full: full)

        var n = try count(&r)
        var last: UInt64? = nil
        for _ in 0..<n {
            let id = try nextId(&r, &last)
            entities[id] = nil
            summary.despawned.append(id)
        }

        n = try count(&r)
        last = nil
        for _ in 0..<n {
            let id = try nextId(&r, &last)
            let q = [try r.zigzag(), try r.zigzag(), try r.zigzag()]
            var components: [UInt8: Data] = [:]
            try readComponents(&r, &components)
            entities[id] = ReplicatedEntity(id: id, q: q, position: (0, 0, 0), components: components)
            summary.spawned.append(id)
        }

        n = try count(&r)
        last = nil
        for _ in 0..<n {
            let id = try nextId(&r, &last)
            let mask = try r.u8()
            guard var e = entities[id] else {
                throw ReplicationError.malformed("update for unknown entity \(id)")
            }
            for (i, bit) in [UInt8(1), 2, 4].enumerated() where mask & bit != 0 {
                // Wrapping, like the encoder: exact over the whole Int64 range.
                e.q[i] = e.q[i] &+ (try r.zigzag())
            }
            if mask & 8 != 0 { try readComponents(&r, &e.components) }
            entities[id] = e
            summary.updated.append(id)
        }
        guard r.done else { throw ReplicationError.malformed("trailing bytes") }

        let p = Double(precision)
        for (id, var e) in entities {
            e.position = (Double(e.q[0]) * p, Double(e.q[1]) * p, Double(e.q[2]) * p)
            entities[id] = e
        }
        return summary
    }

    private func count(_ r: inout Reader) throws -> Int {
        let n = try r.varint()
        guard n <= Self.maxCount else { throw ReplicationError.malformed("count too large") }
        return Int(n)
    }

    private func nextId(_ r: inout Reader, _ last: inout UInt64?) throws -> UInt64 {
        let v = try r.varint()
        let id: UInt64
        if let prev = last {
            let (sum, overflow) = prev.addingReportingOverflow(v)
            guard v > 0, !overflow else { throw ReplicationError.malformed("ids do not ascend") }
            id = sum
        } else {
            id = v
        }
        last = id
        return id
    }

    private func readComponents(_ r: inout Reader, _ into: inout [UInt8: Data]) throws {
        let n = try r.varint()
        guard n <= 256 else { throw ReplicationError.malformed("more than 256 components") }
        for _ in 0..<Int(n) {
            let id = try r.u8()
            let len = try r.varint()
            if len == 0 {
                into[id] = nil
                continue
            }
            into[id] = try r.take(len - 1)
        }
    }
}

private struct Reader {
    let bytes: [UInt8]
    var offset = 0

    var done: Bool { offset >= bytes.count }

    mutating func u8() throws -> UInt8 {
        guard offset < bytes.count else { throw ReplicationError.malformed("ends early") }
        defer { offset += 1 }
        return bytes[offset]
    }

    mutating func f32() throws -> Float {
        guard offset + 4 <= bytes.count else { throw ReplicationError.malformed("ends early") }
        var bits: UInt32 = 0
        for i in 0..<4 { bits |= UInt32(bytes[offset + i]) << (8 * UInt32(i)) }
        offset += 4
        return Float(bitPattern: bits)
    }

    mutating func varint() throws -> UInt64 {
        var v: UInt64 = 0
        for i in 0..<10 {
            let b = try u8()
            let part = UInt64(b & 0x7f)
            if i == 9 && part > 1 { throw ReplicationError.malformed("varint past 64 bits") }
            v |= part << UInt64(7 * i)
            if b & 0x80 == 0 { return v }
        }
        throw ReplicationError.malformed("varint too long")
    }

    mutating func zigzag() throws -> Int64 {
        let z = try varint()
        return Int64(bitPattern: z >> 1) ^ -Int64(bitPattern: z & 1)
    }

    mutating func take(_ n: UInt64) throws -> Data {
        guard n <= UInt64(bytes.count - offset) else {
            throw ReplicationError.malformed("component runs past the end")
        }
        let end = offset + Int(n)
        defer { offset = end }
        return Data(bytes[offset..<end])
    }
}

File: packages/realtime/src/replication.ts
/**
 * Entity replication frames (see `pylon_replication::frame` in Rust).
 *
 * A shard that replicates entities sends each subscriber, per tick, the
 * entities that left its view (despawn), came into view (spawn, full
 * state), or changed (update: only the axes that moved, as a quantized
 * difference, and the components that changed or were removed).
 * {@link EntityTable} applies them; after each frame it holds what the
 * server says this subscriber sees.
 *
 * ```text
 * u8      version (1)
 * u8      flags: bit 0 FULL: clear the table first
 * f32 LE  precision
 * varint  despawn count, ids (first absolute, then differences; ascending)
 * varint  spawn count, per entity: id, x, y, z (zigzag, quantized), components
 * varint  update count, per entity: id, u8 mask (1 x, 2 y, 4 z, 8 components),
 *         changed axes (zigzag differences), components if bit 8
 * components: varint count, per component: u8 id, varint (length + 1),
 *         bytes; length field 0 means removed
 * ```
 *
 * JavaScript numbers hold integers exactly up to 2^53. Entity ids and
 * quantized positions must stay below that; a frame that exceeds it is
 * refused rather than rounded.
 */

export const REPLICATION_VERSION = 1;

const FLAG_FULL = 1;
const MASK_X = 1;
const MASK_Y = 2;
const MASK_Z = 4;
const MASK_COMPONENTS = 8;

/** Longest id list a frame may declare. */
const MAX_COUNT = 1 << 24;

export class ReplicationError extends Error {
  constructor(message: string) {
    super(`replication frame: ${message}`);
    this.name = "ReplicationError";
  }
}

/** One entity as the client sees it. */
export interface ReplicatedEntity {
  readonly id: number;
  /** Quantized position; `x/y/z` are these times the frame precision. */
  qx: number;
  qy: number;
  qz: number;
  x: number;
  y: number;
  z: number;
  /** Component id to bytes, in the game's own encoding. */
  components: Map<number, Uint8Array>;
}

/** What one frame did. */
export interface ReplicationSummary {
  full: boolean;
  spawned: number[];
  updated: number[];
  despawned: number[];
}

class Reader {
  private offset = 0;
  private readonly view: DataView;

  constructor(private readonly bytes: Uint8Array) {
    this.view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  }

  get done(): boolean {
    return this.offset >= this.bytes.length;
  }

  u8(): number {
    if (this.offset >= this.bytes.length) throw new ReplicationError("ends early");
    return this.bytes[this.offset++];
  }

  f32(): number {
    if (this.offset + 4 > this.bytes.length) throw new ReplicationError("ends early");
    const v = this.view.getFloat32(this.offset, true);
    this.offset += 4;
    return v;
  }

  /** An unsigned LEB128 varint, as a bigint (exact to 64 bits). */
  varint(): bigint {
    let v = 0n;
    for (let i = 0; i < 10; i++) {
      const b = this.u8();
      const part = BigInt(b & 0x7f);
      if (i === 9 && part > 1n) throw new ReplicationError("varint past 64 bits");
      v |= part << BigInt(7 * i);
      if ((b & 0x80) === 0) return v;
    }
    throw new ReplicationError("varint too long");
  }

  /** A zigzag signed varint, as a bigint. */
  zigzag(): bigint {
    const z = this.varint();
    return (z >> 1n) ^ -(z & 1n);
  }

  take(n: number): Uint8Array {
    if (this.offset + n > this.bytes.length) throw new ReplicationError("component runs past the end");
    const out = this.bytes.slice(this.offset, this.offset + n);
    this.offset += n;
    return out;
  }
}

function toSafe(v: bigint, what: string): number {
  if (v > BigInt(Number.MAX_SAFE_INTEGER) || v < BigInt(Number.MIN_SAFE_INTEGER)) {
    throw new ReplicationError(`${what} ${v} is beyond 2^53`);
  }
  return Number(v);
}

function count(r: Reader): number {
  const n = r.varint();
  if (n > BigInt(MAX_COUNT)) throw new ReplicationError("count too large");
  return Number(n);
}

function nextId(r: Reader, last: number | null): number {
  const v = r.varint();
  if (last === null) return toSafe(v, "entity id");
  if (v === 0n) throw new ReplicationError("ids do not ascend");
  return toSafe(BigInt(last) + v, "entity id");
}

function readComponents(r: Reader, into: Map<number, Uint8Array>): void {
  const n = r.varint();
  if (n > 256n) throw new ReplicationError("more than 256 components");
  for (let i = 0; i < Number(n); i++) {
    const id = r.u8();
    const len = r.varint();
    if (len === 0n) {
      into.delete(id);
      continue;
    }
    if (len - 1n > BigInt(Number.MAX_SAFE_INTEGER)) throw new ReplicationError("component too large");
    into.set(id, r.take(Number(len - 1n)));
  }
}

/**
 * The entities one subscriber has been told about. Apply every
 * replication frame in order; a frame that fails to apply means the
 * connection is out of sync and should be reopened (the server then sends
 * a full frame).
 */
export class EntityTable {
  readonly entities = new Map<number, ReplicatedEntity>();
  /** World units per quantization step, from the last frame. */
  precision = 0.01;

  get size(): number {
    return this.entities.size;
  }

  get(id: number): ReplicatedEntity | undefined {
    return this.entities.get(id);
  }

  clear(): void {
    this.entities.clear();
  }

  apply(frame: Uint8Array): ReplicationSummary {
    const r = new Reader(frame);
    const version = r.u8();
    if (version !== REPLICATION_VERSION) throw new ReplicationError(`version ${version}`);
    const full = (r.u8() & FLAG_FULL) !== 0;
    const precision = r.f32();
    if (!Number.isFinite(precision) || precision <= 0) throw new ReplicationError("bad precision");
    if (full) this.entities.clear();
    this.precision = precision;
    const summary: ReplicationSummary = { full, spawned: [], updated: [], despawned: [] };

    let n = count(r);
    let last: number | null = null;
    for (let i = 0; i < n; i++) {
      last = nextId(r, last);
      this.entities.delete(last);
      summary.despawned.push(last);
    }

    n = count(r);
    last = null;
    for (let i = 0; i < n; i++) {
      last = nextId(r, last);
      const qx = toSafe(r.zigzag(), "position");
      const qy = toSafe(r.zigzag(), "position");
      const qz = toSafe(r.zigzag(), "position");
      const components = new Map<number, Uint8Array>();
      readComponents(r, components);
      this.entities.set(last, {
        id: last,
        qx,
        qy,
        qz,
        x: qx * precision,
        y: qy * precision,
        z: qz * precision,
        components,
      });
      summary.spawned.push(last);
    }

    n = count(r);
    last = null;
    for (let i = 0; i < n; i++) {
      last = nextId(r, last);
      const mask = r.u8();
      const e = this.entities.get(last);
      if (!e) throw new ReplicationError(`update for unknown entity ${last}`);
      if (mask & MASK_X) e.qx = toSafe(BigInt(e.qx) + r.zigzag(), "position");
      if (mask & MASK_Y) e.qy = toSafe(BigInt(e.qy) + r.zigzag(), "position");
      if (mask & MASK_Z) e.qz = toSafe(BigInt(e.qz) + r.zigzag(), "position");
      if (mask & MASK_COMPONENTS) readComponents(r, e.components);
      summary.updated.push(last);
    }
    if (!r.done) throw new ReplicationError("trailing bytes");

    // Positions follow the frame's precision (a full frame may change it).
    for (const e of this.entities.values()) {
      e.x = e.qx * precision;
      e.y = e.qy * precision;
      e.z = e.qz * precision;
    }
    return summary;
  }
}
